odilo.py

```diff
@@ -0,0 +1,598 @@
+    def _do_post(url, payload, headers, **kwargs):
+        # More time please
+        if 'timeout' not in kwargs:
+            kwargs['timeout'] = 60
+
+        return HTTP.post_with_timeout(url, payload, headers=headers, **kwargs)
+
+
+class MockOdiloAPI(OdiloAPI):
+    @classmethod
+    def mock_collection(cls, _db):
+        library = DatabaseTest.make_default_library(_db)
+        collection, ignore = get_one_or_create(
+            _db, Collection,
+            name="Test Odilo Collection",
+            create_method_kwargs=dict(
+                external_account_id=u'library_id_123',
+            )
+        )
+        integration = collection.create_external_integration(
+            protocol=ExternalIntegration.ODILO
+        )
+        integration.username = u'username'
+        integration.password = u'password'
+        integration.setting(OdiloAPI.LIBRARY_API_BASE_URL).value = u'http://localhost:8080/api/v2'
+        library.collections.append(collection)
+
+        return collection
+
+    def __init__(self, _db, collection, *args, **kwargs):
+        self.access_token_requests = []
+        self.requests = []
+        self.responses = []
+
+        self.access_token_response = self.mock_access_token_response('token')
+        super(MockOdiloAPI, self).__init__(_db, collection, *args, **kwargs)
+
+    def token_post(self, url, payload, headers={}, **kwargs):
+        """Mock the request for an OAuth token.
+        """
+
+        self.access_token_requests.append((url, payload, headers, kwargs))
+        response = self.access_token_response
+        return HTTP._process_response(url, response, **kwargs)
+
+    def mock_access_token_response(self, credential):
+        token = dict(token=credential, expiresIn=3600)
+        return MockRequestsResponse(200, {}, json.dumps(token))
+
+    def queue_response(self, status_code, headers={}, content=None):
+        self.responses.insert(
+            0, MockRequestsResponse(status_code, headers, content)
+        )
+
+    def _do_get(self, url, *args, **kwargs):
+        """Simulate Representation.simple_http_get."""
+        response = self._make_request(url, *args, **kwargs)
+        return response.status_code, response.headers, response.content
+
+    def _do_post(self, url, *args, **kwargs):
+        return self._make_request(url, *args, **kwargs)
+
+    def _make_request(self, url, *args, **kwargs):
+        response = self.responses.pop()
+        self.requests.append((url, args, kwargs))
+        return HTTP._process_response(
+            url, response, kwargs.get('allowed_response_codes'),
+            kwargs.get('disallowed_response_codes')
+        )
+
+
+class OdiloRepresentationExtractor(object):
+    """Extract useful information from Odilo's JSON representations."""
+
+    log = logging.getLogger("OdiloRepresentationExtractor")
+
+    format_data_for_odilo_format = {
+        "ACSM": (
+            Representation.PDF_MEDIA_TYPE, DeliveryMechanism.ADOBE_DRM
+        ),
```

Comment:
Do you know that the media type here is PDF, or might it be EPUB or something else?

Reply:
ACSM resources could be either PDF or EPUB, so, I'am not sure about what representation media type should we use here, since in addition the actual files received are .acsm (which internally are xml-like).

Reply:
DeliveryMechanism.ADOBE_DRM describes the ACSM file. The first item in this tuple conveys what type of document you get _after_ fulfilling the ACSM file. It's important to convey this information because some clients might not support every format. For example, the SimplyE app currently supports EPUB but not PDF.

We don't have a way to accommodate a situation where the content provider just doesn't know what the media type is. Even if we could accommodate this in the database, OPDS doesn't have a way to represent that you might get one media type or the other and there's no way to know without trying it. Even if we could represent that in OPDS, it wouldn't be very useful because the client needs to know whether it can actually render a book before offering the user the chance to borrow it.

I can think of four resolutions:

1. Set the first media type to Representation.PDF_MEDIA_TYPE. Your books are either PDF or EPUB, and SimplyE (the primary client that will be rendering these books) supports EPUB but not PDF. Once SimplyE supports PDF, we'll know that we can render all the books you offer through this mechanism, and they'll start showing up in the mobile client. We'll need to make sure SimplyE can handle a case where we thought we were getting PDF but we actually get EPUB, but as long as the document itself is served with the correct media type it should be okay.
2. Set the first media type to Representation.EPUB_MEDIA_TYPE. This will make your collection's titles show up in SimplyE immediately, but if a book turns out to be a PDF, there will be an unexplained rendering error when SimplyE actually fulfills the ACSM file and tries to render the document. If the vast majority of your titles are actually EPUBs, this might be good enough as a stopgap until we add PDF support to SimplyE.
3. Set the first media type in the 2-tuple as `application/octet-stream;type=https://bib.schema.org/EBook`. `application/octet-stream` is the most general media type and using it indicates that you don't know the 'real' media type. This has the advantage of always being accurate, but clients are generally not comfortable handling arbitrary binary data.
However, https://www.iana.org/assignments/media-types/application/octet-stream allows for the media type to be given a "type" parameter indicating "the general type or category of binary data". The type is not intended to be processed automatically, but that recommendation was written in 1996 before we had a set of well-known URIs describing every general type of thing that might be delivered over the Internet. My idea is to use the schema.org URI for an ebook (https://bib.schema.org/EBook) to indicate that the document is some kind of ebook. This will signal to a client that's confident it can handle _most_ types of ebooks to say to the user "I can _probably_ render this book but I won't know for sure until you borrow it and I'm able to download it."
We can then change SimplyE to understand `application/octet-stream;type=https://bib.schema.org/EBook` and to give that warning. If the book happens to be an EPUB, then SimplyE will be able to render it. If it's a PDF, or some other unknown data type, then SimplyE will fail to render the book and give an error that makes sense in terms of that initial warning.
4. If you're absolutely certain that PDF and EPUB are the only two things that can be behind an "ACSM" type in your system, we can do solution 3 but use something like `http://odilo.us/terms/epub-or-pdf` as the `type` parameter instead of `https://bib.schema.org/EBook`. This will let us get rid of the warning altogether once we add PDF support to SimplyE.

Reply:
I always forget to mention this. Even after SimplyE adds PDF support, we will not be able to render PDF documents that we obtained through fulfilling an ACSM file. This is a limitation imposed by our contract with Adobe, not a technical limitation. So even with solution #4, SimplyE will always need to issue a warning that it might not be able to open the book after the user borrows it.